src/NServiceBus.Core.Tests/API/TaskReturning.cs

```diff
@@ -0,0 +1,287 @@
+﻿namespace NServiceBus.Core.Tests.API
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+    using System.Reflection;
+    using System.Runtime.CompilerServices;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using NUnit.Framework;
+
+    static class TaskReturning
+    {
+        static readonly List<MethodInfo> allMethods = typeof(IMessage).Assembly.GetTypes()
+            .Where(type => !type.IsObsolete())
+            .SelectMany(type => type.GetMethods(BindingFlags.Static | BindingFlags.Instance | BindingFlags.DeclaredOnly | BindingFlags.Public | BindingFlags.NonPublic))
+            .Where(method => !method.IsCompilerGenerated())
+            .Where(method => !method.IsObsolete())
+            .ToList();
+
+        static readonly List<MethodInfo> taskMethods = allMethods
+            .Where(method => typeof(Task).IsAssignableFrom(method.ReturnType))
+            .ToList();
+
+#pragma warning disable IDE0001 // Simplify Names
+        static readonly List<MethodInfo> noTokenPolicy = taskMethods
+            .Where(method =>
+                (typeof(Delegate).IsAssignableFrom(method.DeclaringType) && method.Name == "EndInvoke") ||
+                (method.HasCancellableContext() &&
+                    method.IsOnOneOf(
+                        typeof(NServiceBus.Saga),
+                        typeof(NServiceBus.IncomingMessageOperations),
+                        typeof(NServiceBus.IPipeline),
+                        typeof(NServiceBus.Pipeline.IBehavior),
+                        typeof(NServiceBus.IHandleMessages<>),
+                        typeof(NServiceBus.IHandleTimeouts<>),
+                        typeof(NServiceBus.Sagas.IHandleSagaNotFound),
+                        typeof(NServiceBus.Pipeline.MessageHandler),
+                        typeof(NServiceBus.IUnicastPublishRouter)
+                        )) ||
+                method.IsOn(typeof(NServiceBus.IAsyncTimer)) ||
+                method.IsOn(typeof(NServiceBus.ICancellableContext)) ||
+                method.IsOn(typeof(NServiceBus.DelayedMessagePoller)) ||
+                (method.IsOn(typeof(NServiceBus.TaskEx)) && method.Name == "ThrowIfNull") || // TODO: Consider removing this type?
+                method.IsOn(typeof(NServiceBus.MessageMutator.IMutateIncomingMessages)) ||
+                method.IsOn(typeof(NServiceBus.MessageMutator.IMutateIncomingTransportMessages)) ||
+                method.IsOn(typeof(NServiceBus.MessageMutator.IMutateOutgoingMessages)) ||
+                method.IsOn(typeof(NServiceBus.MessageMutator.IMutateOutgoingTransportMessages))
```

Comment:
The context objects that the methods on these types operate on do not inherit from `ICancellableContext`, which leads me to believe that either they should, or these methods should accept a token.

Am I missing something?

Reply:
@adamralph These got missed because mutators aren't a bona-fide part of the pipeline, they're kind of a detour from the pipeline. I think we assumed the contexts had the same ancestry that now includes ICancellableContext. I'm updating now.